### DISCUSSION TITLE
Rules are complex, and we want aggregation and real-time computation rules to support logical operations

### DISCUSSION THREAD

**Discussion by wego-xc** [prometheus/prometheus#13284] [Q&A]
### Proposal

Rules are complex, and we want aggregation and real-time computation rules to support logical operations

I hope the rule engine is officially supported
Or just use a lua script, which is great

**Comment by beorn7** [comment]
Not sure what you are up to here. Could you be more specific?

Not the following:
- Prometheus supports recording rules: https://prometheus.io/docs/prometheus/latest/configuration/recording_rules/
- PromQL supports logical operations: https://prometheus.io/docs/prometheus/latest/querying/operators/
- Adding scripting language support was prototyped in https://github.com/prometheus/prometheus/pull/12851 and discussed multiple times at [dev-summits](https://docs.google.com/document/d/11LC3wJcVk00l8w5P3oLQ-m3Y37iom6INAMEu2ZAGIIE/edit?pli=1#bookmark=kix.b6g7wjdfisa5), with the result so far that we are hesitant to support a scripting language engine for now.

**Comment by wego-xc** [comment]
> Not sure what you are up to here. Could you be more specific?
> 
> Not the following:
> 
> * Prometheus supports recording rules: https://prometheus.io/docs/prometheus/latest/configuration/recording_rules/
> * PromQL supports logical operations: https://prometheus.io/docs/prometheus/latest/querying/operators/
> * Adding scripting language support was prototyped in [Add User Defined Functions to PromQL #12851](https://github.com/prometheus/prometheus/pull/12851) and discussed multiple times at [dev-summits](https://docs.google.com/document/d/11LC3wJcVk00l8w5P3oLQ-m3Y37iom6INAMEu2ZAGIIE/edit?pli=1#bookmark=kix.b6g7wjdfisa5), with the result so far that we are hesitant to support a scripting language engine for now.

We want the logical operation to refer to if... else or support for scripting language engines can complement this if...else problem.

**Comment by beorn7** [comment]
PromQL is not an imperative language. It's more like a functional language. The `or` operator gives you something like "if/else", quoting the documentation: "`vector1 or vector2` results in a vector that contains all original elements (label sets + values) of `vector1` and additionally all elements of `vector2` which do not have matching label sets in `vector1`."

**Comment by wego-xc** [comment]
> PromQL is not an imperative language. It's more like a functional language. The `or` operator gives you something like "if/else", quoting the documentation: "`vector1 or vector2` results in a vector that contains all original elements (label sets + values) of `vector1` and additionally all elements of `vector2` which do not have matching label sets in `vector1`."

But how do we solve a similar problem

**Comment by beorn7** [comment]
I'll convert this into a discussion. If you describe your problem in more detail, perhaps some people can help you in the discussion.

The feature request for an embedded scripting language is valid, of course, but we discussed it at one of the recent dev-summits and decided to not pursue it for now. We left #12851 open, though, as a reminder where to pick it up, should we reconsider.
